Add block/unblock/remove APIs. (#24)
riscv-non-isa/riscv-ap-tee: specification/sbi_atee.adoc

Reviewer: general comment - the "block" terminology may be confusing - I have been using "invalidate page mapping" and "validate page mapping" terms - perhaps we can use the same for these ABI functions.

Author: Adding @abrestic-rivos to provide his input, as he designed these APIs.

Reviewer 2: I'm fine with the rename. My only complaint with the "invalidate/validate page mappings" is it's a bit verbose. How about `{in,}validate_ptes` or `{in,}validate_pages` or `set_ptes_{present,invalid}`, etc.

Author: `{in,}validate_pages` SGTM.

TVMs can choose to yield access to confidential memory at runtime and request shared (non-confidential) memory.
The TVM must communicate its request to the host to convert confidential to non-confidential and vice-versa 
explicitly via the `sbi_tee_guest_share_memory_region()` and `sbi_tee_guest_unshare_memory_region()`. This request
results in an exit to the TSM which enforces the security properties on the mapping and exits to the VMM host.
If the region of address space is populated, the host must first invalidate and remove the confidential pages.
This requires the host to make three separate ECALLs to the TSM:
 

> . `sbi_tee_host_tvm_block_pages()`

Reviewer: ```suggestion
. `sbi_tee_host_tvm_invalidate_page_maps()`
```

| SBI_ERR_FAILED          | The operation failed for unknown reasons.
|===

> [#sbi_tee_host_tvm_block_pages]

Reviewer: ```suggestion
[#sbi_tee_host_tvm_invalidate_page_maps]
```

> === Function: TEE Host TVM Block Pages (FID #16)

Reviewer: ```suggestion
=== Function: TEE Host TVM Invalidate Page Mappings (FID #16)
```

[source, C]
-----

> struct sbiret sbi_tee_host_tvm_block_pages(unsigned long tvm_guest_id,

Reviewer: ```suggestion
struct sbiret sbi_tee_host_tvm_invalidate_page_maps(unsigned long tvm_guest_id,
```

                                           unsigned long gpa,
                                           unsigned long length);
-----

> Invalidates the pages in the specified range of guest physical address space and

Reviewer: ```suggestion
Invalidates the page mappings in the specified range of guest physical address space and
```

> marks the pages as blocked.

Reviewer: ```suggestion
thus marks the pages as blocked from creating new cached mappings.
```

For each page in the range, the TSM must verify that:

* The page is currently marked present in the TVM’s page table.
* The page is either mapped and uniquely owned by the TVM, or shared and owned by the host.

After verifying these pre-conditions are met, the TSM then invalidates the pages and

> marks the pages as blocked. The host must complete a TVM TLB invalidation sequence,

Reviewer: ```suggestion
marks the pages as not-present. The host must complete a TVM TLB invalidation sequence,
```

initiated by `sbi_tee_host_tvm_fence()`, in order to complete the invalidation.

> Guest page faults taken by the TVM on blocked pages continue to be reported to the host.

Reviewer: ```suggestion
Guest page faults taken by the TVM on invalid page mappings continue to be reported to the host.
```

> The page remains invalid until the mapping is unblocked (marked present) removed, or

Reviewer: ```suggestion
The page mapping remains invalid until the mapping is unblocked (marked present) removed, or
```

becomes part of a huge page by promotion/demotion operation.

The possible error codes returned in `sbiret.error` are shown below.

> [#table_sbi_tee_host_tvm_block_pages_errors]

Reviewer: ```suggestion
[#table_sbi_tee_host_tvm_invalidate_page_maps_errors]
```

> .TEE Host TVM Block Pages

Reviewer: ```suggestion
.TEE Host TVM Invalidate Page Mappings
```

[cols="2,3", width=90%, align="center", options="header"]
|===
| Error code              | Description
| SBI_SUCCESS             | The operation completed successfully.
| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `length` were invalid.
| SBI_ERR_INVALID_ADDRESS | `gpa` was invalid.
| SBI_ERR_FAILED          | The operation failed for unknown reasons.
|===

> [#sbi_tee_host_tvm_unblock_pages]

Reviewer: can you please make similar changes for the unblock_pages -> validate_page_maps
